privacy/scrubber.go

```diff
@@ -81,6 +82,63 @@ func NewScrubber() Scrubber {
 	return scrubber{}
 }
 
+func (scrubber) ScrubRequest(bidRequest *openrtb2.BidRequest, enforcement Enforcement) *openrtb2.BidRequest {
+	if enforcement.UFPD {
+		// transmitUfpd covers user.ext.data, user.data, user.id, user.buyeruid, user.yob, user.gender, user.keywords, user.kwarray
+		// and device.{ifa, macsha1, macmd5, dpidsha1, dpidmd5, didsha1, didmd5}
+		if bidRequest.Device != nil {
+			bidRequest.Device.DIDMD5 = ""
+			bidRequest.Device.DIDSHA1 = ""
+			bidRequest.Device.DPIDMD5 = ""
+			bidRequest.Device.DPIDSHA1 = ""
+			bidRequest.Device.IFA = ""
+			bidRequest.Device.MACMD5 = ""
+			bidRequest.Device.MACSHA1 = ""
+		}
+		if bidRequest.User != nil {
+			bidRequest.User.Data = nil
+			bidRequest.User.Ext = scrubExtIDs(bidRequest.User.Ext, "data")
+			bidRequest.User.ID = ""
+			bidRequest.User.BuyerUID = ""
+			bidRequest.User.Yob = 0
+			bidRequest.User.Gender = ""
+			bidRequest.User.Keywords = ""
+			bidRequest.User.KwArray = nil
+		}
+	}
+	if enforcement.Eids {
+		//transmitEids covers user.eids and user.ext.eids
+		if bidRequest.User != nil {
+			bidRequest.User.EIDs = nil
+			bidRequest.User.Ext = scrubExtIDs(bidRequest.User.Ext, "eids")
+		}
+	}
+
+	if enforcement.TID {
+		//remove source.tid and imp.ext.tid
+		if bidRequest.Source != nil {
+			bidRequest.Source.TID = ""
+		}
+		for ind, imp := range bidRequest.Imp {
+			impExt := scrubExtIDs(imp.Ext, "tid")
+			bidRequest.Imp[ind].Ext = impExt
+		}
+	}
+
+	if enforcement.PreciseGeo {
+		//round user's geographic location by rounding off IP address and lat/lng data.
+		//this applies to both device.geo and user.geo
+		if bidRequest.User != nil && bidRequest.User.Geo != nil {
+			bidRequest.User.Geo = scrubGeoPrecision(bidRequest.User.Geo)
+		}
+		if bidRequest.Device != nil && bidRequest.Device.Geo != nil {
+			bidRequest.Device.Geo = scrubGeoPrecision(bidRequest.Device.Geo)
+		}
+	}
```

Comment:
Does `scrubGeoPrecision` also round off the IPv4 and IPv6 IP addresses?

Reply:
Wow, I totally missed this part! Good catch! 

For the bidRequest.Device.IPv6  should we use `scrubIPV6Lowest32Bits` or `scrubIPV6Lowest16Bits`? 
Current code:
```go
	bidRequest.Device.IPv6 = scrubIPV6Lowest32Bits(bidRequest.Device.IPv6)
```

Reply:
Actually, we should've made that configurable a long time ago. Seems we missed that. 

For now, think it's likely Lowest32Bits is a good choice. What does PBS-Java use as their default? I

Reply:
In PBS Java here is a function to scrub IpV6: https://github.com/prebid/prebid-server-java/blob/3ebefb8ccdb52e68bb604660fc9808b54a513536/src/main/java/org/prebid/server/auction/IpAddressHelper.java#L39 and here is a simple unit tests for it: https://github.com/prebid/prebid-server-java/blob/3ebefb8ccdb52e68bb604660fc9808b54a513536/src/test/java/org/prebid/server/auction/IpAddressHelperTest.java#L65

It works a little different from what we have in Go:
For the Ip = "1111:2222:3333:4444:5555:6666:7777:8888" func scrubIPV6Lowest32Bits will return "1111:2222:3333:4444:5555:6666:0:0". It makes sense, lowest 32 bits were changed to 0s.

In Ipv6 lowest 64 bits represent physical or device Mac address and upper 64 bits represent a network address. 
Upper 48 bits represent global address and next 16 bits represent subnet id. 
http://www.steves-internet-guide.com/ipv6-guide/
https://en.wikipedia.org/wiki/IPv6_address
It looks like PBS Javaonly leaves a global address and scrub subnet id and device Mac address.

Are there any requirements to check what is the right way to scrub IpV6?